### PR TITLE
[build_templates] [hostcfgd] Keep containers hostname up to date

### PR DESCRIPTION
**Pull request by msosyak**
**- What I did**
Implemented logic to keep containers hostname up to date

**- How I did it**
* Added updateHostName function to docker_image_ctl.j2
* Added hostname specification on container creating step
* Added listener for hostname changes in hostcfgd which call updateHostName for all running containers if an appropriate script(generated from docker_image_ctl.j2) exist in /usr/bin 

**- How to verify it**
Change hostname in config_db.json
Load new configuration
Check hostname in containers

**- Description for the changelog**
Added logic to keep containers hostname up to date

@akokhan @pyadvichuk @vsenchyshyn @qiluo-msft @jleveque @lguohan

### CODE REVIEW

File: files/image_config/hostcfgd/hostcfgd [hunk 1]
```diff
@@ -166,10 +167,40 @@ class HostConfigDaemon:
             log_data['passkey'] = obfuscate(log_data['passkey'])
         syslog.syslog(syslog.LOG_INFO, 'value of {} changed to {}'.format(key, log_data))
 
+    def hostname_handler(self, key, data):
```
Comment (qiluo-msft) [inline files/image_config/hostcfgd/hostcfgd]: >key [](start = 31, length = 3)

Check key == localhost?

File: files/image_config/hostcfgd/hostcfgd [hunk 2]
```diff
@@ -166,10 +167,40 @@ class HostConfigDaemon:
...
+        hostname = data["hostname"]
+        if hostname == self.hostname_cache:
+            return
+
+        syslog.syslog(syslog.LOG_INFO, "Get all running containers")
+        cmd = 'docker ps --format "{{.Names}}"'
+        try:
+            containers = subprocess.check_output(cmd, shell=True).split("\n")[:-1]
+        except subprocess.CalledProcessError as err:
+                syslog.syslog(syslog.LOG_WARNING, "{} - failed: return code - {}, output:\n{}"
```
Comment (qiluo-msft) [inline files/image_config/hostcfgd/hostcfgd]: >                 [](start = 0, length = 16)

Inconsistent indentation

File: files/image_config/hostcfgd/hostcfgd [hunk 3]
```diff
@@ -166,10 +167,40 @@ class HostConfigDaemon:
...
+                                .format(err.cmd, err.returncode, err.output))
+
+        for name in containers:
+            script = '/usr/bin/{}.sh'.format(name)
+            exists = os.path.isfile(script)
+            if not exists:
+                syslog.syslog(syslog.LOG_INFO, "Can't find controll script for {}".format(name))
```
Comment (qiluo-msft) [inline files/image_config/hostcfgd/hostcfgd]: >LOG_INFO [](start = 37, length = 8)

This is an error.

File: files/image_config/hostcfgd/hostcfgd [hunk 4]
```diff
@@ -166,10 +167,40 @@ class HostConfigDaemon:
...
+                continue
+
+            cmd = "{} updateHostName {}".format(script, hostname)
```
Comment (qiluo-msft) [inline files/image_config/hostcfgd/hostcfgd]: >hostname [](start = 56, length = 8)

Possible to inject attack from Redis, such as 'rm -rf /'

File: files/image_config/hostcfgd/hostcfgd (hunk 2 above)
Comment (qiluo-msft) [inline files/image_config/hostcfgd/hostcfgd]: >LOG_WARNING [](start = 37, length = 11)

This is an error.

File: files/image_config/hostcfgd/hostcfgd [hunk 5]
```diff
@@ -166,10 +167,40 @@ class HostConfigDaemon:
...
+            try:
+                subprocess.check_call(cmd, shell=True)
+            except subprocess.CalledProcessError as err:
+                syslog.syslog(syslog.LOG_WARNING, "{} - failed: return code - {}, output:\n{}"
```
Comment (qiluo-msft) [inline files/image_config/hostcfgd/hostcfgd]: >LOG_WARNING [](start = 37, length = 11)

This is an error.

File: files/build_templates/docker_image_ctl.j2 [hunk 6]
```diff
@@ -5,6 +5,28 @@ function getMountPoint()
...
+
+    # copy HOSTS to HOSTS_TMP
+    $EXEC "cp $HOSTS $HOSTS_TMP"
+    # remove entry with hostname
+    $EXEC "sed -i \"/$HOSTNAME$/d\" $HOSTS_TMP"
+    # add entry with new hostname
+    $EXEC "echo -e \"127.0.0.1\t$NEW_HOSTNAME\" >> $HOSTS_TMP"
+
+    echo "Set hostname in {{docker_container_name}} container"
+    $EXEC "hostname $NEW_HOSTNAME"
```
Comment (qiluo-msft) [inline files/build_templates/docker_image_ctl.j2]: >NEW_HOSTNAME [](start = 21, length = 12)

Possible to input injection

File: files/build_templates/docker_image_ctl.j2 [hunk 7]
```diff
@@ -153,6 +180,7 @@ start() {
 {%- endif %}
         --tmpfs /tmp \
         --tmpfs /var/tmp \
+        --hostname $HOSTNAME \
```
Comment (qiluo-msft) [inline files/build_templates/docker_image_ctl.j2]: >$HOSTNAME [](start = 19, length = 9)

Possible to input injection

File: files/image_config/hostcfgd/hostcfgd (hunk 1 above)
Comment (myronsosyak) [reply files/image_config/hostcfgd/hostcfgd]: Added validation

File: files/image_config/hostcfgd/hostcfgd (hunk 4 above)
Comment (myronsosyak) [reply files/image_config/hostcfgd/hostcfgd]: Added validation

**Comment by myronsosyak** [comment]
Review comments fixed

File: files/build_templates/docker_image_ctl.j2 (hunk 7 above)
Comment (qiluo-msft) [reply files/build_templates/docker_image_ctl.j2]: Still possible. For example:
`
sonichost"; rm -rf; echo "
`

---
In reply to: [286279985](https://github.com/Azure/sonic-buildimage/pull/2924#discussion_r286279985) [](ancestors = 286279985)

Comment (msosyak) [reply files/build_templates/docker_image_ctl.j2]: Bash did not process metacharacters from variables only metacharacters directly present in a script or interpreter line.
It is easy to check by the next commands:
```bash
print_args() {
while (( "$#" )); do 
  echo $1 
  shift 
done
}
X="hostName"
Y='sonic"; echo injection; echo "'
print_args --hostname "$X" --name cname
#output
>>> --hostname
hostName
--name
cname

print_args --hostname "$Y" --name cname
#output
>>> --hostname
sonic"; echo injection; echo "
--name
cname
```

I thought that [CVE-2016-0634 ](https://www.cvedetails.com/cve/CVE-2016-0634/)can be exploited there. But sonic use bash 4.4

So, for any case added validation for allowed characters in a hostname.

**Comment by mykolaf** [comment]
Please use a standard commit message format:

>     [component/folder touched]: Description intent of your changes
>
>     [List of changes]
>
> 	  Signed-off-by: Your Name your@email.com

File: files/build_templates/docker_image_ctl.j2 (hunk 6 above)
Comment (msosyak) [reply files/build_templates/docker_image_ctl.j2]: Added validation for allowed characters in a hostname.

File: files/build_templates/docker_image_ctl.j2 (hunk 7 above)
Comment (msosyak) [reply files/build_templates/docker_image_ctl.j2]: @qiluo-msft Could we close this conversation?

Comment (qiluo-msft) [reply files/build_templates/docker_image_ctl.j2]: Nice comments! Really appreciate the clarification and reference. 

---
In reply to: [286898693](https://github.com/Azure/sonic-buildimage/pull/2924#discussion_r286898693) [](ancestors = 286898693)

File: files/build_templates/docker_image_ctl.j2 [hunk 8]
```diff
@@ -5,6 +5,31 @@ function getMountPoint()
...
+    # copy HOSTS to HOSTS_TMP
+    $EXEC "cp $HOSTS $HOSTS_TMP"
+    # remove entry with hostname
+    $EXEC "sed -i \"/$HOSTNAME$/d\" $HOSTS_TMP"
+    # add entry with new hostname
+    $EXEC "echo -e \"127.0.0.1\t$NEW_HOSTNAME\" >> $HOSTS_TMP"
+
+    echo "Set hostname in {{docker_container_name}} container"
+    $EXEC "hostname '$NEW_HOSTNAME'"
+    $EXEC "cat $HOSTS_TMP > $HOSTS"
```
Comment (jleveque) [inline files/build_templates/docker_image_ctl.j2]: Suggest to add `$EXEC "rm -f $HOSTS_TMP"`

Comment (msosyak) [reply files/build_templates/docker_image_ctl.j2]: Added

File: files/image_config/hostcfgd/hostcfgd [hunk 9]
```diff
@@ -166,10 +177,49 @@ class HostConfigDaemon:
...
+            containers = subprocess.check_output(cmd, shell=True).split("\n")[:-1]
+        except subprocess.CalledProcessError as err:
+            syslog.syslog(syslog.LOG_ERR, "{} - failed: return code - {}, output:\n{}"
+                          .format(err.cmd, err.returncode, err.output))
+
+        for name in containers:
+            script = '/usr/bin/{}.sh'.format(name)
+            exists = os.path.isfile(script)
+            if not exists:
+                syslog.syslog(syslog.LOG_ERR, "Can't find controll script for {}".format(name))
```
Comment (jleveque) [inline files/image_config/hostcfgd/hostcfgd]: Typo: s/controll/control

Comment (msosyak) [reply files/image_config/hostcfgd/hostcfgd]: Fixed

**Comment by akokhan** [comment]
@jleveque , @lguohan ,
as I see, all review comments are resolved now. Is there anything else that blocks us from merging this PR?

**Comment by jleveque** [comment]
Retest this please

**Pull request merged**